cookbook/en/making-http-requests.md

```diff
@@ -0,0 +1,367 @@
+
+```shell
+composer require guzzlehttp/psr7 --prefer-dist
+```
+
+## Basic usage
+
+### Simple GET request
+
+Here's how to make a basic GET request using PSR-18 interfaces:
+
+```php
+<?php
+
+declare(strict_types=1);
+
+use GuzzleHttp\Client;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+
+class ApiService
+{
+    public function __construct(
+        private ClientInterface $httpClient,
+        private RequestFactoryInterface $requestFactory,
+    ) {
+    }
+
+    public function fetchUserData(int $userId): ResponseInterface
+    {
+        $request = $this->requestFactory->createRequest(
+            'GET',
+            "https://jsonplaceholder.typicode.com/users/{$userId}"
+        );
+
+        return $this->httpClient->sendRequest($request);
+    }
+}
+```
+
+### POST request with JSON data
+
+Here's an example of making a POST request with JSON payload:
+
+```php
+<?php
+
+declare(strict_types=1);
+
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+
+class UserService
+{
+    public function __construct(
+        private ClientInterface $httpClient,
+        private RequestFactoryInterface $requestFactory,
+        private StreamFactoryInterface $streamFactory,
+    ) {
+    }
+
+    public function createUser(array $userData): ResponseInterface
+    {
+        $jsonData = json_encode($userData, JSON_THROW_ON_ERROR);
+        $stream = $this->streamFactory->createStream($jsonData);
+
+        $request = $this->requestFactory->createRequest('POST', 'https://jsonplaceholder.typicode.com/users')
+            ->withHeader('Content-Type', 'application/json')
+            ->withHeader('Accept', 'application/json')
+            ->withBody($stream);
+
+        return $this->httpClient->sendRequest($request);
+    }
+}
+```
+
+## Configuration in Yii3
+
+### Container configuration
+
+Configure the HTTP client and PSR factories in your DI container:
+
+```php
+<?php
+
+declare(strict_types=1);
+
+// config/common/http-client.php
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Psr7\HttpFactory;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+
+return [
+    ClientInterface::class => [
+        'class' => Client::class,
+        '__construct()' => [
+            'config' => [
+                'timeout' => 30,
+                'connect_timeout' => 10,
+            ],
+        ],
+    ],
+    
+    RequestFactoryInterface::class => HttpFactory::class,
+    ResponseFactoryInterface::class => HttpFactory::class,
+    StreamFactoryInterface::class => HttpFactory::class,
+    UriFactoryInterface::class => HttpFactory::class,
```

Comment:
This is problematic since there could be non-guzzle implementation of PSR-7 used already. Is there a better way?